Add support for background colors, and bold and underlined font

martinvonz opened the pull request:
<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the branch rather than adding commits on top. Use force-push when
pushing the updated branch (`jj git push` does that automatically when you
rewrite a branch). Merge the PR at will once it's been approved. See
https://github.com/martinvonz/jj/blob/main/docs/contributing.md for details.
-->

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have added tests to cover my changes

dbarnett reviewed (approved):
Awesome! Clear readability improvement and crossterm was already in dependencies.

dbarnett commented:
Looks like there are some unnecessary `.clone()`s clippy wants removed and some insta snapshots that changed slightly, though.

martinvonz commented:
> Looks like there are some unnecessary `.clone()`s clippy wants removed and some insta snapshots that changed slightly, though.

I was working on unit tests for the formatters anyway, so I think I'll make sure to include snapshot tests of the current color codes in those. Then I'll rebase this commit on top of the tests. So it'll take a little while before I update this PR, but I hope it'll make us more confident in it.

dbarnett reviewed (approved):
Now it looks good *and* passes tests!

ilyagr reviewed (commented):
Not that it matters, but I think the discussion you're referring to in the commit description wasn't on Discord, but rather in https://github.com/martinvonz/jj/pull/941#issuecomment-1363561554.

martinvonz commented:
> Not that it matters, but I think the discussion you're referring to in the commit description wasn't on Discord, but rather in [#941 (comment)](https://github.com/martinvonz/jj/pull/941#issuecomment-1363561554).

Thanks. I've updated the comment.

ilyagr commented:
I think your color correspondence is correct according to https://docs.rs/crossterm/latest/src/crossterm/style/types/color.rs.html#125. 

I think it's good to stick with "bright black" and such colors in jj's UI (as you did). I find crossterm color names weird. In particular, correct me if I'm wrong, but I think that on Linux, the bright bit makes the color brighter *and* makes the test bold on many terminals (the one I have on hand right now is xterm.js). So, text marked "DarkGrey" will always be bold since it's actually brightblack. Also, these color names are library-dependent. In Go's https://github.com/gdamore/tcell, [`Grey` is brightblack and `Silver` is white](https://github.com/gdamore/tcell/blob/3e6ca936791efc42a380c44760cb2d527864d4d0/color.go#L60).

ilyagr commented on src/formatter.rs:
```diff
@@ -291,22 +292,22 @@ mod tests {
-        [1;35mbright magenta[0m
-        [1;36mbright cyan[0m
-        [1;37mbright white[0m
+        [38;5;0mblack[39m
```
Do you know why this test changed? What is the difference between `1;30` and `38;5;1`? Why did `0m` become `39m`?

I might look up the answer myself, but I mainly wanted to double-check that this is expected.

martinvonz commented:
> I think your color correspondence is correct according to https://docs.rs/crossterm/latest/src/crossterm/style/types/color.rs.html#125 (perhaps it's worth inserting a comment about this in the code). **Update:** I see the tests double-check this, which makes inserting a comment less important, but perhaps still useful. Up to you.

Thanks for checking. Yes, I think the tests make it clear enough. I added the tests after writing the initial commit description (and thus the PR description).

> 
> I think it's good to stick with "bright black" and such colors in jj's UI (as you did). I find crossterm color names weird. In particular, correct me if I'm wrong, but I think that on Linux, the bright bit makes the color brighter _and_ makes the test bold on many terminals (the one I have on hand right now is xterm.js). So, text marked "DarkGrey" will always be bold since it's actually brightblack. Also, these color names are library-dependent. In Go's https://github.com/gdamore/tcell, [`Grey` is brightblack and `Silver` is white](https://github.com/gdamore/tcell/blob/3e6ca936791efc42a380c44760cb2d527864d4d0/color.go#L60).

Using escape code 1 for "bold" does often result in bold text *and* bright color, IIUC. That seems to be the case for in iTerm2. Setting the bright color directly with escape code 38 (as after this PR) does *not* result in also getting bold text in iTerm2. I had not noticed that before, but now that I look carefully, it does mean that the working-copy commit is less obvious in the log output. So maybe I shouldn't merge this PR until I've added support for bolding.

martinvonz commented:
This is now ready for review again (sorry I forgot to make it draft while it wasn't). The scope grew to not just be about crossterm but to also add support for bold and underlined fonts and for background color.

yuja commented on src/formatter.rs:
```diff
@@ -184,26 +197,28 @@ impl<W: Write> ColorFormatter<W> {
     fn write_new_color(&mut self) -> io::Result<()> {
-        let new_color = self.current_color();
-        if new_color != self.current_color {
+        let new_style = self.current_style();
```
Nit: also rename this function to `write_new_style()`?

yuja commented on src/formatter.rs:
```diff
@@ -137,43 +138,38 @@ impl Default for Style {
 pub struct ColorFormatter<W> {
     output: W,
-    colors: Arc<HashMap<String, String>>,
+    rules: Arc<HashMap<Vec<String>, Style>>,
```
We can also change `HashMap` to `Vec`, but that's not required here.

yuja commented on src/formatter.rs:
```diff
@@ -156,36 +162,35 @@ impl<W: Write> ColorFormatter<W> {
+            // "a b c", we'll get [2,0]. We compare them by Rust's default vector
+            // comparison. That means "a c" will trump both rule "b" (priority [1]) and rule
+            // "a b" (priority [1,0]).
+            let mut matched_styles = vec![];
```
Nit: it would be nice if "c" (priority [2]) is included in the example. "a c" vs "a b" only requires the maximum index, but "a c" vs "c" doesn't.

martinvonz replied:
Good point. I think I meant something different than I wrote. I've update the example to be more complex and hopefully more useful.

yuja commented on src/formatter.rs:
```diff
@@ -211,6 +213,18 @@ impl<W: Write> ColorFormatter<W> {
+                    // our record of the current style so we re-apply the other attributes
+                    // below.
+                    queue!(self.output, SetAttribute(Attribute::Reset))?;
+                    self.current_style = Style::default();
```
I don't know which is better, but I think it's also fine to always reset and apply merged attributes and colors.

martinvonz replied:
I'm also not sure, but I've made some other changes to make us emit fewer escapes. I'll close this, but let me know if you think I should change anything.

yuja commented on src/config/colors.toml:
```diff
@@ -48,6 +44,7 @@
 "op-log user" = "yellow"
 "op-log time" = "cyan"
 "op-log tags" = "white"
+"op-log head" = "bold"
```
`{ bold = true }` ?

ilyagr commented on tests/test_global_opts.rs:
```diff
@@ -156,8 +156,8 @@ color="always""#,
     );
     let stdout = test_env.jj_cmd_success(&repo_path, &["log", "-T", "commit_id"]);
     insta::assert_snapshot!(stdout, @r###"
-    @ [1;34m230dd059e1b059aefc0da06a2e5a7dbf22362f22[0m
```
This output changed slightly (the color is unchanged).

![image](https://user-images.githubusercontent.com/4123047/211689430-b0d36da7-db36-4281-9fb5-ff4214fcc4de.png)

martinvonz replied:
Yeah, that depends on your terminal. Which one are you using? I think I tried on iTerm2 and xterm and they both rendered the old format as bold and bright even though we only said that it should be bold. Your terminal seems to make "bold" only bold. Since it's so common for terminals to render bold as bold+bright, it felt like a regression to lose the bright colors on those terminals.

By the way, I have noticed another regression in this PR, which is that there's new an unwanted blank line after the for line of the description in the working-copy commit. I'll look into that later, but expect an update to this PR.

ilyagr replied:
All the other tests render correctly. I think there's something (very mildly) wrong here. I see it in both Kitty and Gnome Terminal.

It is correct for terminals to render text with the bold attribute as bright, I think. However, some terminal color themes use the same color for bright as for non-bright, making the text just bold.

By the way, my current favorite way of looking at these is to add `.diff` to the github's commit URL, run the `sponge | less -R` command from `moreutils` in the terminal, paste the content there, and press `^D`. If you have these files on disk, you can just run `less -R` on your files.

martinvonz replied:
Which commit is that diff from? Whichever it is, the old output is not bright but the new output is bright. My guess is that you're looking at the commit that started using `crossterm`. From that commit, we explicitly set bright color and bold font.

ilyagr replied:
Actually, I have no reason to claim this is wrong. Just weird. Whatever was wrong is actually fixed by this commit. Not sure what's going on or whether it's worth investigating.  So, never mind.

ilyagr replied:
Didn't see your response before last comment. I think we're in agreement. Yes, I was looking at the `crossterm` commit.

ilyagr replied:
There's a discussion about that: https://codeberg.org/dnkl/foot/issues/199. According to the author, "terminals are moving away from drawing bold text as bright", though it's not true for any of the ones I use.

spectral54 commented on src/formatter.rs:
```diff
@@ -408,18 +424,41 @@ mod tests {
+        insta::assert_snapshot!(String::from_utf8(output).unwrap(), @r###"
+        [38;5;1m fg only [39m
+        [48;5;4m bg only [49m
+        [38;5;2m[48;5;3m single rule [39m[49m
```
Can we add a test for `"bright_red_fg white_bg"(bright red on white)"blue_fg <unset_bg>"(dark blue on white)`? Because I think the bg will be reset/lost in that case.

spectral54 replied:
(Though if all of these commits go in at the same time I'm ok with skipping this test, since we're removing the hack later... or maybe include the test now, but label it as broken, and remove that label later when it works?)

martinvonz replied:
Good catch. I've added the test case. That actually also caught a bug in the underlining commit. I had accidentally added a `self.current_style = Style::default();` line there too, which led to us not resetting the foreground color and background color after removing underlining and other attributes at the same time.

spectral54 commented on src/formatter.rs:
```diff
@@ -430,7 +450,8 @@ mod tests {
             r#"
         colors.fg = { fg = "red" }
         colors.bg = { bg = "blue" }
```
Same here, can we use a different label than 'bg'? And in the future commit, 'bold'. :)

ilyagr commented on src/formatter.rs:
```diff
@@ -408,18 +424,41 @@ mod tests {
         formatter.write_str(" fg only ").unwrap();
         formatter.remove_label().unwrap();
-        insta::assert_snapshot!(String::from_utf8(output).unwrap(), @"[38;5;1m fg only [39m");
+        formatter.write_str("\n").unwrap();
```
It would be nice to have a test somewhere that if `colors."abra cadabra" = "green"`, a single label doesn't make it green and both labels together do, in either order.

ilyagr replied:
Thank you! That's nice documentation.

> We actually require the order to match since https://github.com/martinvonz/jj/pull/993.

I see, I missed most of that discussion.

ilyagr commented on src/config/colors.toml:
```diff
@@ -20,11 +20,7 @@
 
-# TODO: This near-duplication of the lines above is unfortunate. Should we
-# allow adding and clearing the "bright" bit somehow? Or should we instead
-# use a different background color? (We don't have support for background
```
Perhaps this TODO should change to making it possible to set "bright" an attribute like "bold" and "underlined"? You could have a separate "bright_bg" trait, but it would probably be better to simply let people use "bright" both as an attribute or as a color name.

Or do you think it's not worth the trouble?

martinvonz replied:
Since "bright" only applies to the 8 base colors, I don't think we should do it. It would make it complicated when we add support for more than the 8 base colors.

ilyagr replied:
Good point. I'm not sure if `jj` should ever use more than 16 colors but we shouldn't restrict ourselves now.

martinvonz commented on src/formatter.rs:
```diff
@@ -294,22 +319,22 @@ mod tests {
+        [38;5;5m magenta [39m
+        [38;5;6m cyan [39m
+        [38;5;7m white [39m
+        [1m[38;5;8m bright black [0m[39m
```
Looking at https://github.com/crossterm-rs/crossterm/blob/master/src/command.rs, it seems to only care on Windows. Maybe we should check that ourselves and disable color? I'll leave that for later either way.

martinvonz commented on src/formatter.rs:
```diff
@@ -294,22 +319,22 @@ mod tests {
+        [38;5;5m magenta [39m
+        [38;5;6m cyan [39m
+        [38;5;7m white [39m
+        [1m[38;5;8m bright black [0m[39m
```
It doesn't seem to have any effect. I feel like it shouldn't affect the `ColorFormatter` anyway, but we could add a check in the application logic to create a `PlainTextFormatter` instead.

martinvonz commented on src/formatter.rs:
```diff
@@ -294,22 +319,22 @@ mod tests {
+        [38;5;5m magenta [39m
+        [38;5;6m cyan [39m
+        [38;5;7m white [39m
+        [1m[38;5;8m bright black [0m[39m
```
We do: https://github.com/martinvonz/jj/blob/e6cdd69d834705eb7c94a0adc24f957388a3d09c/src/config.rs#L134-L138

martinvonz commented on src/formatter.rs:
```diff
@@ -294,22 +319,22 @@ mod tests {
+        [38;5;7m white [39m
+        [1m[38;5;8m bright black [0m[39m
+        [1m[38;5;9m bright red [0m[39m
+        [1m[38;5;10m bright green [0m[39m
```
Fixed in a later commit now

yuja commented on src/formatter.rs:
```diff
@@ -176,6 +176,11 @@ pub struct ColorFormatter<W> {
+    /// written to the terminal. We don't write the style to the terminal
+    /// until we actually write some text. This is just an optimization; there
+    /// should be no functional difference if this is always `true`.
+    has_pending_style_change: bool,
```
I think this flag is redundant since we have `current_style` (actual) and `requested_style()` (expected).

This might serve as an optimization to avoid `cached_styles` lookup, but I don't know if that would matter.

martinvonz replied:
You're probably right. I guess it would mostly matter if we printed lots of text without changing style (so this flag would let us skip the check), but we rarely do that, so it probably doesn't matter. I'll drop the flag.

yuja commented on src/formatter.rs:
```diff
@@ -353,10 +353,26 @@ fn color_for_name(color_name: &str) -> Option<Color> {
+                self.output.write_all(&line[..line.len() - 1])?;
+                let labels = self.labels.clone();
+                self.labels.clear();
+                self.write_new_style()?;
```
`mem::take()` or `mem::replace()` can be used to temporarily reset labels, but I think it would be better to pass `labels` as an argument.

martinvonz replied:
I went with `mem::tale()` and `mem::replace()` because `write_new_style()` mutably borrows `self`, so we would need to pass a clone of the labels otherwise. Let me know if there was a way that I just missed.

yuja replied:
Right. Maybe we would have to break things further to pass labels as argument.

Nit: `mem::replace()` without using old value is just an assignment.

martinvonz replied:
D'oh. That's why it wanted me to use the result. Fixed :)

yuja commented on src/formatter.rs:
```diff
@@ -353,10 +353,26 @@ fn color_for_name(color_name: &str) -> Option<Color> {
     fn write(&mut self, data: &[u8]) -> Result<usize, Error> {
-        if self.has_pending_style_change {
-            self.write_new_style()?;
+        for line in data.split_inclusive(|b| *b == b'\n') {
```
Maybe better to copy the commit message as inline comment? It'll help understand why we have to split lines and write one by one.

martinvonz replied:
Good idea. Will do.

martinvonz replied:
Oops, I forgot to actually do that. Working on it.

martinvonz commented:
Thanks for reviewing, everyone! I've enabled auto-merge on this, so this should go in soon. If there are any further comments, I'll address them in a followup.

martinvonz commented on src/formatter.rs:
```diff
@@ -242,3 +242,220 @@ impl<W: Write> Formatter for ColorFormatter<W> {
+        formatter.remove_label().unwrap();
+        formatter.write_str(" b2 ").unwrap();
+        formatter.remove_label().unwrap();
+        formatter.write_str(" a2 ").unwrap();
```
I just heard about https://lists.apache.org/thread/dz20fzls5mw8l6zlv0nmnsv6tl1qyk0t, which talks about security implications of printing unsanitized inputs (including inputs from the repo). I think we should sanitize in `ColorFormatter`. We'll just need to make sure that `jj print`[^1] prints its output to a `PlainTextFormatter` even if color was requested. Or maybe it's better to add a `write_raw/write_unescaped/write_unsafe` method to the `Formatter` trait.

[^1]: Should we rename that to `jj cat` to match `hg cat`, btw?

ilyagr replied:
I'm no expert, but it seems to me that PlainTextFormatter should also sanitize, even if that just means removing all the terminal escapes (it's probably better to replace them them like `less` without `-R` does). I think PlainTextFormatter-ed strings get printed to the terminal occasionally, and even if that's not currently the case, it would be counter-intuitive if that was more dangerous than using ColorFormatter.

So, `write_unsafe` seems to me like the way to go.

If we wanted to be serious about this, we would have to check every `println!` or `write!` inside `jj` for whether it can print untrusted data to somewhere other than a sanitizing Formatter, whether that can show up in the terminal, and if so whether that's OK for some exceptional reason (like maybe `jj print`).

Also , +1 on renaming `jj print` to `jj cat`.

martinvonz replied:
> it seems to me that PlainTextFormatter should also sanitize

Yes, that seems obvious now that you say it :)

> would have to check every `println!` or `write!`

We always write to a `Formatter` that we get from the `Ui` object, so that should already be safe. There are just a few exceptions where we print messages from library code (e.g. "ignoring git submodule foo" or "upgrading to protobuf").

yuja replied:
I think `jj cat`'s behavior should be switched depending on isatty(stdout). If tty, remove weird escape sequences. Otherwise dump the raw content (to e.g. a file.)

So maybe SanitizedFormatter or something, than `write_unsafe()`.

martinvonz replied:
Ah, that's a very good point. So you're saying we create a `SanitizedFormatter` that's wraps a `Formatter` and implements `Formatter` itself? Sounds good.

yuja replied:
or make both Sanitized/ColorFormatter go through the sanitization pass? My point is we'll need dynamic dispatch between sanitized and raw outputs.

ilyagr replied:
@yuja 's idea is clever, but the topic of security brings out a Debbie Downer in me. I feel like it's dangerous for security features to be too context-dependent. Would we really want `jj cat` to be secure while `jj cat | cat` isn't? It seems safer to have `jj cat` not be too smart and have an `--insecure` flag or something.

yuja replied:
I consider it as a feature (to help CLI user), not a real security issue. `cat` should dump file content unmodified, but if the output is a tty, user wouldn't care the exact content, so preprocessing is allowed.

ilyagr replied:
It makes a lot of sense -- from that perspective, there's not really a downside. It's very possible that `jj cat` isn't a big concern from the security perspective, in which case the user-friendly perspective is the only one that matters.
